include/shell.h

```diff
@@ -55,7 +55,7 @@ class ByteReader {
 
 class HostShell {
 public:
-	virtual bool GetEnvStr(const char* entry, std::string& result) const = 0;
+	[[nodiscard]] virtual std::optional<std::string> GetEnvStr(std::string_view entry) const = 0;
```

Comment:
Let's discard the `[[nodiscard]]`s with prejudice 😎

Reply:
`discard the [[nodiscard]]`,  :rofl: 

The thing it tells me is that I need to handle and do something with the return value, otherwise calling the function does nothing of value.

```c++
(void) GetEnvStr("testing");
```

is 100% equivalent to doing nothing, and therefore the compiler will give you an error if you try the above, where as with other functions, it'll happily carry on.

Reply:
But we need to do something with the return value in all functions that return a value? So where is the value here? We could add `[[nodiscard]]` to every function and I don't see what would improve. It's basically noise.

Reply:
Ok, I guess you're saying @kcgen that for pure functions that have no side effects to start using this... Well, I still find it very much superfluous; it adds visual noise, but if I don't do anything with the return value, surely my code that uses the function won't work... Dead easy to spot those "errors", but I don't even think I ever made such an error to use a pure function that returns a value then not do anything with the value.

Seems like a solution to a non-issue to me 🤷🏻

Reply:
To elaborate further, if I look at `GetEnvStr`, it's obvious it returns a value, and most people would assume it does not mutate anything. I don't see how `[[nodiscard]]` makes anyones life better by chucking before "get" functions, adding visual noise. Now, if it _did_ mutate anything in this "get" function, probably it should be renamed and/or documented. Like `GetEnvStrAndCheckEmail` or dunno... you get my drift 😎

Reply:
Okay, so what's your long-term plan here; add `[[nodiscard]]` to _every single function_ that returns a value? Because surely no function just returns something "just for the fun of it", right? So why not annotate them all, "just in case" it helps?


Real world scenario:

1. When I write some code that calls some function because **I want a value**, I want to do something with that value one or two lines below—that's why I called that function to get the value in the first place, right?
2. What's the likelihood of me _not realising_ one or two lines below that hey, I forgot to assign that value to a var, a value **that I absolutely need** to do what I want to do? Like... zero? 😄 

Let me simplify it:

1. I want to drink a glass of water. Ok, so now I call `GetGlass()` that gives me a glass (I hold the glass in my right hand at this point).
2. Second, I call `PourWaterIntoGlass()`, and I need to pass in _something_ to that function I want to pour the water into (I extend my right hand with the glass in it and open the tap).

What's the likelihood of me sitting there with a dumb expression on my face "not being able to find the bug" because instead of holding the glass in my hand, I let it fall to the floor, and then I open the tap... and where's my glass of water? 😆  How do I even fix this bug... I don't know, it's just too hard... 😝 

Another example:

```c
FILE* fp = fopen("file.txt", "w+");
fseek(fp, 0, SEEK_SET);
```

How can this feature "help" me if I do this?

```c
fopen("file.txt", "w+");
fseek(fp, 0, SEEK_SET);
```

I get a compilation error and it doesn't take a genius to find the "bug" that I forgot to store `fp`...

I guess if our static analysers flag such discarded return value issues then sure, let's do that. But polluting the codebase with `[[nodiscard]]` on every single function, just because it _might_ help someone is a bit of a crazy idea.

Just because people chuck all sorts of things into newer C++ standards it doesn't mean we all must use them without thinking about it a bit first, or that they are even well-thought-out features.

So yeah it wouldn't be a _completely_ worthless warning, but the benefits are extremely minor as I see it. And we'd pollute our codebase significantly. A case of much ado about... well, very slim _potential_ benefits. So yeah, please feel free to configure the static analysers to flag these cases where a return value is discarded if we don't have it configured already.

Reply:
Thanks for your input on this @MeAreJeenius. I think we've already said everything we can about the subject to each other on this one 😄

Ultimately, the maintainers of the project decide on the overall style and what goes in and what doesn't. In this case, I see this a low-value thing that just adds noise, therefore I don't want these annotations to go in, so please remove them. If there was one universal C++ style to follow, there wouldn't be so many team-specific styles per every project 😄 

However, like I said, you can look into turning on a compiler warning that would issue warnings without having to add these annotations, or maybe extending the static analyser configs to flag unused return values (if we're not already flagging them; I have no idea).

Regardless of this _small_ thing or disagreement, I very much value your functional contributions in this ticket. Great stuff 👍🏻